File: sandbox/algm/leetcode/binary tree level order traversal/BFS.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
		List<List<Integer>> ret = new LinkedList<List<Integer>>();
		Queue<TreeNode> toVisit = new LinkedList<TreeNode>();
		if(root == null) return(ret);
		toVisit.offer(root);
		while (! toVisit.isEmpty()) {
			int n = toVisit.size();
			List<Integer> curL = new LinkedList<Integer>();
			for (int i = 0; i < n; i++) {
				TreeNode cur = toVisit.poll();			
				if (cur.left != null) toVisit.offer(cur.left);
				if (cur.right != null) toVisit.offer(cur.right);
				curL.add(cur.val);
			}
			ret.add(curL);
		}
		return(ret);
    }
}

File: sandbox/algm/lintcode/binary-tree-level order traversal/BFS.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Level order in a list of lists of integers
    """
    def levelOrder(self, root):
        # write your code here
        #try BFS first
        if root is None:
            return([])
        from Queue import Queue
        q = Queue()
        q.put(root)
        ret = []
        while not q.empty():
            curLvl = []
            n = q.qsize()
            for i in range(n):
                e = q.get()
                curLvl.append(e.val)
                if e.left:
                    q.put(e.left)
                if e.right:
                    q.put(e.right)
            ret.append(curLvl)
        return(ret)
                
            

